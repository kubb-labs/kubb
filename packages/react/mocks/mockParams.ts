import type { Params } from '../src/types.ts'

export const mockParams: Array<{ name: string; params: Params }> = [
  {
    name: 'empty',
    params: {
      name: {},
    },
  },
  {
    name: 'string',
    params: {
      name: {
        type: 'string',
      },
    },
  },
  {
    name: 'optional string',
    params: {
      name: {
        type: 'string',
        optional: true,
      },
    },
  },
  {
    name: 'default string',
    params: {
      name: {
        type: 'string',
        default: 'test',
      },
    },
  },
  {
    name: 'undefined',
    params: {
      name: undefined,
    },
  },
  {
    name: 'name and address',
    params: {
      name: {
        type: 'string',
      },
      address: {
        type: 'string',
      },
    },
  },
  {
    name: 'name[optional], age[optional] and address',
    params: {
      name: {
        type: 'string',
        optional: true,
      },
      age: {
        type: 'number',
        optional: true,
      },
      address: {
        type: 'string',
      },
    },
  },
  {
    name: 'name[optional] age, address[optional] and country[Belgium]',
    params: {
      name: {
        type: 'string',
        optional: true,
      },
      age: {
        type: 'number',
      },
      address: {
        type: 'string',
        optional: true,
      },
      country: {
        type: 'string',
        default: '"Belgium"',
      },
    },
  },
  {
    name: 'params[with name] with age, address[optional] and country',
    params: {
      data: {
        type: 'Data',
        children: {
          age: {
            type: 'number',
          },
          address: {
            type: 'string',
            optional: true,
          },
          country: {
            type: 'string',
          },
        },
      },
    },
  },
  {
    name: 'params[without name] with age, address[optional] and country',
    params: {
      data: {
        children: {
          age: {
            type: 'number',
          },
          address: {
            type: 'string',
            optional: true,
          },
          country: {
            type: 'string',
          },
        },
      },
    },
  },
  {
    name: 'undefined children param',
    params: {
      data: {
        children: {
          age: undefined,
        },
      },
    },
  },
  {
    name: 'empty children param',
    params: {
      data: {
        children: {
          age: {},
        },
      },
    },
  },
  {
    name: 'params[without name and default] with age[optional, 90] address[optional] and country[optional, Belgium]',
    params: {
      data: {
        default: '{}',
        children: {
          age: {
            type: 'number',
            default: '90',
            optional: true,
          },
          address: {
            type: 'string',
            optional: true,
          },
          country: {
            type: 'string',
            default: "'Belgium'",
            optional: true,
          },
        },
      },
    },
  },
  {
    name: 'params[without name] with age[optional, 20] address[optional] and country[optional, Belgium]',
    params: {
      data: {
        children: {
          age: {
            type: 'number',
            default: '20',
            optional: true,
          },
          address: {
            type: 'string',
            optional: true,
          },
          country: {
            type: 'string',
            default: '"Belgium"',
            optional: true,
          },
        },
      },
    },
  },
  {
    name: 'params[without name and mode inline] with age[optional, 20] address[optional] and country[optional, Belgium]',
    params: {
      data: {
        mode: 'inline',
        children: {
          age: {
            type: 'number',
            default: '20',
            optional: true,
          },
          address: {
            type: 'string',
            optional: true,
          },
          country: {
            type: 'string',
            default: '"Belgium"',
            optional: true,
          },
        },
      },
    },
  },
  {
    name: 'params[without name and mode inline] with age[20] address and country[Belgium]',
    params: {
      data: {
        mode: 'inline',
        children: {
          age: {
            type: 'number',
            default: '20',
          },
          address: {
            type: 'string',
          },
          country: {
            type: 'string',
            default: '"Belgium"',
          },
        },
      },
    },
  },
  {
    name: 'params[without name and mode inlineSpread] with age[20] address and country[Belgium]',
    params: {
      data: {
        mode: 'inlineSpread',
        children: {
          age: {
            type: 'number',
            default: '20',
          },
          address: {
            type: 'string',
          },
          country: {
            type: 'string',
            default: '"Belgium"',
          },
        },
      },
    },
  },
]
