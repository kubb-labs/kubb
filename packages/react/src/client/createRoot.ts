import { createNode } from '../shared/dom.ts'
import { ReactTemplate } from '../shared/ReactTemplate.tsx'

import type { Logger } from '@kubb/core/logger'
import type { AppContextProps } from '../components/App.tsx'
import type { DOMElement } from '../types.ts'
import type { RootType } from './types.ts'

const instances = new Map<string, ReactTemplate>()

type Props = {
  container?: DOMElement
  logger?: Logger
}

export function createRoot<Context extends AppContextProps = AppContextProps>({ container, logger }: Props = {}): RootType<Context> {
  if (!container) {
    container = createNode('kubb-root')
  }

  const instance = new ReactTemplate<Context>(container, { logger })
  instances.set(instance.id, instance)

  return {
    render(children, context?: Context) {
      return instance.render(children, context)
    },
    unmount() {
      instance.unmount()
      instances.delete(instance.id)
    },
    get output() {
      return instance.output
    },
    get files() {
      return instance.files
    },
    getFile(id: string) {
      return instance.files.find((file) => file.id === id)
    },
  }
}
