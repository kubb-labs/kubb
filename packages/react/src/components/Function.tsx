import { createJSDocBlockText } from '@kubb/core/transformers'

import { Text } from './Text.tsx'

import type { JSDoc, KubbNode } from '../types.ts'

type Props = {
  /**
   * Name of the function.
   */
  name: string
  /**
   * Parameters/options/props that need to be used.
   */
  params?: string
  /**
   * Does this function need to be exported.
   */
  export?: boolean
  /**
   * Does the function has async/promise behaviour.
   * This will also add `Promise<returnType>` as the returnType.
   */
  async?: boolean
  /**
   * Generics that needs to be added for TypeScript.
   */
  generics?: string | string[]

  /**
   * ReturnType(see async for adding Promise type).
   */
  returnType?: string
  /**
   * Options for JSdocs.
   */
  JSDoc?: JSDoc
  children?: KubbNode
}

export function Function({ name, export: canExport, async, generics, params, returnType, JSDoc, children }: Props): KubbNode {
  return (
    <>
      {JSDoc?.comments && (
        <>
          {createJSDocBlockText({ comments: JSDoc?.comments })}
          <br />
        </>
      )}
      {canExport && (
        <Text>
          export
          <Text.Space />
        </Text>
      )}
      {async && (
        <Text>
          async
          <Text.Space />
        </Text>
      )}
      <Text>
        function {name}
        <Text.Space />
      </Text>
      {generics && (
        <>
          <Text>{'<'}</Text>
          <Text>{Array.isArray(generics) ? generics.join(', ').trim() : generics}</Text>
          <Text>{'>'}</Text>
        </>
      )}
      <Text>({params})</Text>
      {returnType && !async && <Text>: {returnType}</Text>}
      {returnType && async && (
        <Text>
          : Promise{'<'}
          {returnType}
          {'>'}
        </Text>
      )}
      <Text>{' {'}</Text>
      <br />
      <Text indentSize={4}>{children}</Text>
      <br />
      <Text>{'}'}</Text>
      <br />
    </>
  )
}

type ArrowFunctionProps = Props & {
  /**
   * Create Arrow function in one line
   */
  singleLine?: boolean
}

export function ArrowFunction({ name, export: canExport, async, generics, params, returnType, JSDoc, singleLine, children }: ArrowFunctionProps): KubbNode {
  return (
    <>
      {JSDoc?.comments && (
        <>
          {createJSDocBlockText({ comments: JSDoc?.comments })}
          <br />
        </>
      )}
      {canExport && (
        <Text>
          export
          <Text.Space />
        </Text>
      )}
      <Text>
        const {name} =<Text.Space />
      </Text>
      {async && (
        <Text>
          async
          <Text.Space />
        </Text>
      )}
      {generics && (
        <>
          <Text>{'<'}</Text>
          <Text>{Array.isArray(generics) ? generics.join(', ').trim() : generics}</Text>
          <Text>{'>'}</Text>
        </>
      )}
      <Text>({params})</Text>
      {returnType && !async && <Text>: {returnType}</Text>}
      {returnType && async && (
        <Text>
          : Promise{'<'}
          {returnType}
          {'>'}
        </Text>
      )}
      {singleLine && (
        <>
          <Text>{' => '}</Text>
          <Text indentSize={4}>{children}</Text>
          <br />
        </>
      )}

      {!singleLine && (
        <>
          <Text>{' => {'}</Text>
          <br />
          <Text indentSize={4}>{children}</Text>
          <br />
          <Text>{'}'}</Text>
          <br />
        </>
      )}
    </>
  )
}

Function.Arrow = ArrowFunction

export const Fun = Function
