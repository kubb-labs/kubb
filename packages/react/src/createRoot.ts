import { ReactTemplate, type ReactTemplateOptions } from './ReactTemplate.tsx'

import type { RootContextProps } from './components/Root.tsx'

export function createRoot<Context extends RootContextProps = RootContextProps>(options: ReactTemplateOptions = { debug: false }): ReactTemplate<Context> {
  return new ReactTemplate<Context>(options)
}
