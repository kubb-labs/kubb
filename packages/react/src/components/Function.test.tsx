import { mockParams } from '../../mocks/mockParams.ts'
import { createRoot } from '../createRoot.ts'
import { Function } from './Function.tsx'

describe('<Function/>', () => {
  test('render Function', async () => {
    const Component = () => {
      return (
        <Function name="getData" export async>
          return 2;
        </Function>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(root.output).toMatchSnapshot()
  })

  test('render Function with comments', async () => {
    const Component = () => {
      return (
        <Function name="getData" export async JSDoc={{ comments: ['@deprecated'] }}>
          return 2;
        </Function>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(root.output).toMatchSnapshot()
  })

  test('render ArrowFunction', async () => {
    const Component = () => {
      return (
        <Function.Arrow name="getData" export async>
          return 2;
        </Function.Arrow>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(root.output).toMatchSnapshot()
  })

  test('render Function Generics', async () => {
    const Component = () => {
      return (
        <Function name="getData" export async generics={['TData']} returnType="number">
          return 2;
        </Function>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(root.output).toMatchSnapshot()
  })

  test('render ArrowFunction Generics', async () => {
    const Component = () => {
      return (
        <Function.Arrow name="getData" export async generics={['TData']} returnType="number">
          return 2;
        </Function.Arrow>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(root.output).toMatchSnapshot()
  })

  test('render ArrowFunction SingleLine', async () => {
    const Component = () => {
      return (
        <Function.Arrow name="getData" export async generics={['TData']} singleLine returnType="number">
          2;
        </Function.Arrow>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(root.output).toMatchSnapshot()
  })

  test('render multiple functions', async () => {
    const Component = () => {
      return (
        <>
          <Function name="getData" export async generics={['TData']} returnType="number">
            2;
          </Function>

          <Function name="getData" export async generics={['TData']} returnType="number">
            3;
          </Function>
        </>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(root.output).toMatchSnapshot()
  })
})
