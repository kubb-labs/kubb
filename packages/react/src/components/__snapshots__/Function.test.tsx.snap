// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Function/> > render ArrowFunction 1`] = `
"export const getData = async () => {
  return 2
}
"
`;

exports[`<Function/> > render ArrowFunction Generics 1`] = `
"export const getData = async <TData,>(): Promise<number> => {
  return 2
}
"
`;

exports[`<Function/> > render ArrowFunction SingleLine 1`] = `
"export const getData = async <TData,>(): Promise<number> => 2
"
`;

exports[`<Function/> > render Function 1`] = `
"export async function getData() {
  return 2
}
"
`;

exports[`<Function/> > render Function Generics 1`] = `
"export async function getData<TData>(): Promise<number> {
  return 2
}
"
`;

exports[`<Function/> > render Function with comments 1`] = `
"/**
 * @deprecated
 */
export async function getData() {
  return 2
}
"
`;

exports[`<Function/> > render multiple functions 1`] = `
"export async function getData<TData>(): Promise<number> {
  2
}
export async function getData<TData>(): Promise<number> {
  3
}
"
`;
